helper/activedirectory/fieldregistry.go

```diff
@@ -0,0 +1,112 @@
+package activedirectory
```

Comment:
Can this file be boiled down to just a few constants (or a single map) that we need?

Reply:
Well, we need a `Parse()` method here https://github.com/hashicorp/vault/pull/4260/files#diff-09726fb0fbf1b5675aa9ad4d9adfe888R17 to be able to handle incoming strings, and it needs to parse it into one of the constants. So that makes it so that just vanilla constants won't quite do it.

It _would_ be possible to take out all the reflection, but then we'd immediately arrive again at a place where the constants need to be added in two places - in an array to help with parsing, and in the constant, which to me leaves an easy window to create a bug when adding new fields in the future.

Reply:
Thanks for linking those two articles, though! I actually hadn't seen them before, and I _do_ have a tendency to do exactly what they say not to do, from a defensive coding perspective. However, now that I know it's not recommended, I can probably immediately revise some of the code like the `activedirectory.Client` to not return an interface, I'll look at that.

Reply:
Hah I meant to add this comment to the next one down......

Reply:
I'll need to think on this one a bit. I think my question is probably more stemming from the introduction of Entry and Field on our side, since the library already has those concepts, exported. But I've not thought enough about the full flow to say whether our versions are/aren't helping.

Reply:
Oh - I mainly did that because I thought it was rather opaque to use the LDAP client for AD to start with. Rather than just pulling out strings and fields, I wanted to start building a knowledge base for future developers using LDAP to ping AD. IOW, a client that makes it easy to have some visibility into what is returned, and has some fields already listed, and speeds future development time.